syntax = "proto3";
package kim;

message addr_info{
    string node_host = 1; /* host for inner server. */
    uint32 node_port = 2; /* port for inner server. */
    string gate_host = 3; /* host for user client. */
    uint32 gate_port = 4; /* port for user client. */
}

message node_info {
    string name = 1;         /* read from config and register to zk. */
    addr_info addr_info = 2; /* network addr info. */
    string node_type = 3;    /* node type in cluster. */
    string conf_path = 4;    /* config path. */
    string work_path = 5;    /* process work path. */
    int32 worker_cnt = 6;    /* number of worker's processes. */
}

/* 
 * zk_node_info
 * manager send to zookeeper, which will be notified to other zk's watchers.
 */
message zk_node {
    string path = 1;        /* zookeeper path. store which return from zk. */
    string type = 2;        /* node type. */
    string host = 3;        /* node host. */
    uint32 port = 4;        /* node port. */
    uint32 worker_cnt = 5;  /* node worker count. */
    double active_time = 6; /* node refresh time. */
}

/* sync zk nodes data betwwen manager and worker. */
message register_node {
    uint32 version = 1;         /* data version. */
    string my_zk_path = 2;      /* my node zk path. */
    repeated zk_node nodes = 3; /* nodes watched.  */
}

/* use in auto send, contact to other node. */
message target_node {
    string node_type = 1;
    string ip = 2;
    uint32 port = 3;
    uint32 worker_index = 4;
}